File: examples/docs-code-snippets/src/main/scala/examples/scala/application/WebShopApp.scala
package examples.scala.application

import cats.effect.IO
import cats.effect.unsafe.implicits.global
import com.ing.baker.compiler.RecipeCompiler
import com.ing.baker.runtime.inmemory.InMemoryBaker
import com.ing.baker.runtime.model.InteractionInstance
import com.ing.baker.runtime.scaladsl.EventInstance
import examples.scala.events.OrderPlaced
import examples.scala.ingredients.Address
import examples.scala.interactions.{CancelOrderImpl, CheckStockImpl, ShipOrderImpl}
import examples.scala.recipes.WebShopRecipe

import java.util.UUID
import scala.concurrent.duration.DurationInt

class WebShopApp {
  def main(args: Array[String]): Unit = {

    val interactions = List(
      new CancelOrderImpl(),
      new ShipOrderImpl(),
      new CheckStockImpl(),
    )

    val bakerF = InMemoryBaker.build(implementations = interactions.map(InteractionInstance.unsafeFrom[IO]))
      .unsafeRunSync()

    val recipeInstanceId = UUID.randomUUID().toString
    val sensoryEvent = EventInstance.unsafeFrom(orderPlaced)

    for {
      recipeId <- bakerF.addRecipe(RecipeCompiler.compileRecipe(recipe = WebShopRecipe.recipe), validate = true)
      _ <- bakerF.bake(recipeId, recipeInstanceId)
      _ <- bakerF.fireSensoryEventAndAwaitReceived(recipeInstanceId, sensoryEvent)
      _ <- bakerF.awaitCompleted(recipeInstanceId, timeout = 5.seconds)
    } yield recipeId
  }

  private val orderPlaced = OrderPlaced(
    orderId = "123",
    customerId = "456",
    productIds = List("iPhone", "PlayStation5"),
    address = Address(
      street = "Hoofdstraat",
      city = "Amsterdam",
      zipCode = "1234AA",
      country = "The Netherlands"
    )
  )
}

File: examples/docs-code-snippets/src/main/java/examples/java/application/WebShopApp.java
package examples.java.application;

import com.ing.baker.compiler.RecipeCompiler;
import com.ing.baker.runtime.inmemory.InMemoryBaker;
import com.ing.baker.runtime.javadsl.EventInstance;
import examples.java.events.OrderPlaced;
import examples.java.ingredients.Address;
import examples.java.interactions.CancelOrderImpl;
import examples.java.interactions.CheckStockImpl;
import examples.java.interactions.ShipOrderImpl;
import examples.java.recipes.WebShopRecipe;

import java.time.Duration;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.ExecutionException;

public class WebShopApp {

    public static void main(String[] args) throws ExecutionException, InterruptedException {
        var baker = InMemoryBaker.java(
            List.of(new CheckStockImpl(), new CancelOrderImpl(), new ShipOrderImpl())
        );

        var recipeInstanceId = UUID.randomUUID().toString();
        var sensoryEvent = EventInstance.from(createOrderPlaced());

        var recipeId = baker.addRecipe(RecipeCompiler.compileRecipe(WebShopRecipe.recipe), true).get();
        baker.bake(recipeId, recipeInstanceId).get();
        baker.fireSensoryEventAndAwaitReceived(recipeInstanceId, sensoryEvent).get();
        baker.awaitCompleted(recipeInstanceId, Duration.ofSeconds(5)).get();
    }

    private static OrderPlaced createOrderPlaced() {
        var address = new Address("Hoofdstraat", "Amsterdam", "1234AA", "The Netherlands");
        return new OrderPlaced("123", "456", address, List.of("iPhone", "Playstation5"));
    }
}
